### ISSUE TITLE
Crashing in NamedQuery when passed js symbol

### ISSUE THREAD

**Issue by darwin** [area1/stpyv8#8]
### The problem:

`CPythonObject::NamedQuery`[1] expects only string names, but it should handle symbol objects as well.

### The solution

This instance of the issue can be trivially fixed by testing for null `*name`. But I'm afraid it will probably require a deeper review of all places where `v8::Name` is used.

Also I would like to open discussion if iterators could be supported. So that PySequence_Check-enabled objects could synthetise this iterator. I will probably need this so I could try to implement it if you ACK the concept.

### The backstory

I started using your library in my pet project [blender-clojure](https://github.com/darwin/blender-clojure). In the beginning I didn't believe I could pull this off, but it works suprisingly well. Good job! 

I got far enough to successfully run non-trivial javascript pieces generated by ClojureScript compiler. And I discovered this issue.

ClojureScript might use code which tries to detect if js object at hand supports [es6 iteration protocol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols). It does it by getting the iterator symbol aka `@@iterator` and doing property lookup on the object with this symbol as name.

It causes a call to `CPythonObject::NamedQuery`[1] with symbol object, it derails and crashes in the call to `::PyObject_HasAttrString`.

I was able to trace it to this exact behaviour:

1. Javascript code does property lookup of `Symbol.iterator`
2. This symbol object gets passed `CPythonObject::NamedQuery`
3. The `v8::String::Utf8Value name(info.GetIsolate(), prop);` fails building the string name and leaves name will null pointer in it (this behaviour is mentioned in V8 include file[2]). It can be also confirmed by looking at the `v8::String::Utf8Value` implementation.
4. The code then calls `::PyObject_HasAttrString` with null pointer as name. This may or may not crash. It reproducibly crashes for me[3] when the Python object type does not implement `tp_getattr`[4], it then follows with `PyUnicode_FromString` with null pointer which is a guaranteed crash [5]

[1] https://github.com/area1/stpyv8/blob/be06d018482db900f5df2bb9d362bad727ad46ac/src/Wrapper.cpp#L386
[2] https://github.com/v8/v8/blob/0c423a7abfdb6e17e997082bc6a0b688d795fcbb/include/v8.h#L3272-L3278
[3]
```
Thread 0:: Dispatch queue: com.apple.main-thread
0   libsystem_platform.dylib      	0x00007fff6b382e52 _platform_strlen + 18
1   org.blenderfoundation.blender 	0x00000001098a3d7e PyUnicode_FromString + 14
2   org.blenderfoundation.blender 	0x000000010987e981 PyObject_GetAttrString + 49
3   org.blenderfoundation.blender 	0x000000010987eb0b PyObject_HasAttrString + 11
4   _STPyV8.cpython-37m-darwin.so 	0x00000001233b01e0 CPythonObject::NamedQuery(v8::Local<v8::Name>, v8::PropertyCallbackInfo<v8::Integer> const&) + 192 (Wrapper.cpp:398)
...
```
[4] https://github.com/python/cpython/blob/6467134307cf01802c9f1c0384d8acbebecbd400/Objects/object.c#L788
[5] https://github.com/python/cpython/blob/6467134307cf01802c9f1c0384d8acbebecbd400/Objects/unicodeobject.c#L2225

**Comment by philsyme** [comment]
Thanks so much for this deep analysis.  Totally agree with your assessment and will ping when we have a candidate fix in.

**Comment by buffer** [comment]
Your analysis is very interesting. Thanks for sharing. I think we should seriously think about supporting iterators so to properly handle this use case too. Please feel free to share samples and patches. Would be really glad to review and contribute.

**Comment by darwin** [comment]
Just FYI. I realized I will need deeper changes for my project. I need to extend wrapper to support CLJS types.

I've got somewhat familiar with V8, Python C/boost API and STPyV8 over the weekend and implemented a proof of concept in https://github.com/darwin/stpyv8.

Originally I thought I would only incrementally add some functionality and keep my fork in sync with yours. Unfortunately I realized I will need to touch the wrapper a bit more. The class hiearachy of js object wrappers is limiting for me. The features I want to support are orthogonal. e.g. a js object instance can be CLJS Type, iterator-like and JS Array at the same time. Mapping it to classes would lead to combinatorial explosion.

I'm going to flatten existing hierarchy it into single CJavascriptObject wrapper and implement all functionality on it - dispatching to different code paths depending on features of the underlying js object. Not sure if this will be a good solution. I'm a bit concerned that having single shared python type could lead to some limitations - e.g. I will have to define some slots and signal to python that they are not implemented when js object won't support it, but that's different than leaving the slot empty. Some slots like `__iter__` let me do that. But some others like `__call__` won't. So maybe I will need several python types anyways. But that will be complexity only in one place (in Expose).

At least that is my current understanding/plan.
